### ISSUE TITLE
[question] why messy code in search result?

### ISSUE THREAD

**Issue by hainfo** [OpenGrok/OpenGrok#972]
I have googled a lot but found not help...
My question is when I index some java file with UTF-8 encoding, once I search the result, I see only messy code in the search result like(&#29723;&#12517 etc). All letters are fine. Do I missed any config?

Thanks in advance!

**Comment by tarzanek** [comment]
yeah, the encoding is all given by locale or by browser, I tried to make most of OpenGrok code UTF-8 compatible, but I might have missed some pieces

If you could attach locale output, sample indexed source file that causes search problems, browser version and browser setup (language wise) and webapp server locale + version/type , then I will try to reproduce and fix any left overs and write more tests o utf-8

I know for sure there are problems with e.g. russian chars in code, it will be the same for chinese, but utf-8 files should hopefully work ... (we do some indexing tests, but we lack UI tests, so there might be hiccups there)

**Comment by tarzanek** [comment]
I meant locale output + env setup for indexer so I know which env is used for indexing (it ideally has to be the same as the webapp servers!)

**Comment by hainfo** [comment]
Thanks for your replay. Finally I found the answer ! 

It am not sure if this is an index issue so I just list the root cause here for your reference...

I got the source file from a jar, and the source file is encoding by UTF-8(with BOM) as I am on windows. Everytime I open the source file with notepad++ it looks good as windows has already taken care of the BOM(byte order char).

I did some experiment:
1. create A.java with UTF8(no BOM) and B.java UTF8(with BOM). Both of these 2 files has some Chinese chars.
2. Open these 2 file with notepad++ and they all looks good.
3. Index these 2 files, and search key word and check result in UI.

The result is that A.java looks good but B.java has messy code.

So the resolution to me is to change the encoding of the source file, but I am not sure if this is an issue for openGrok...

**Comment by tarzanek** [comment]
We should recognize bom
Can you attach both files?
I will try to make them a junit test and fix the webapp

Seems xref is not handling bom well
Might be a restriction of jflex or we just don't pass it as we do to lucene where we take care of bom ...

**Comment by hainfo** [comment]
Seems it does not support any txt/zip file for attachment...
![samplefile](https://cloud.githubusercontent.com/assets/10395563/9076440/d208e1fe-3b54-11e5-8cd0-e8b5615466ce.jpg)
Please download the attached file and rename it to sampleFIle.zip. Hope it woks...

Let me know if this is an issue and if so is there is a plan to resolve this in the future... 
Thanks.

**Comment by tarzanek** [comment]
Hmm you can try to base 64 encode the files and paste them I think this is how github docs suggest to attach small files ...

**Comment by tarzanek** [comment]
e.g. some folks talked about attaching here: https://github.com/ned14/BEurtle/issues/1

**Comment by tarzanek** [comment]
you can pick any base64 encodedr/decoder, I suggest using some command line tool (not sure if windows has one, I hope it should have one)

**Comment by hainfo** [comment]
Thanks for your suggestion.

I cannot access Gist probably due to GFW... So I have attached the test file here:
https://github.com/hainfo/GitHello/blob/master/sampleFile.zip
Hopes it could help.

I also plan to deploy this project on centOS and check the result this weekend.

**Comment by vladak** [comment]
Here's contents of the files in hex if anyone is interested in debugging this further:

```
$ ls -1 | while read file; do echo $file; od -t x1 $file; done
ANSI.java
0000000 74 65 73 74 0d 0a 41 4e 53 49 2e 6a 61 76 61 0d
0000020 0a b2 e2 ca d4 0d 0a d6 d0 ce c4
0000033
UTF8_BOM.java
0000000 ef bb bf 74 65 73 74 0d 0a 55 54 46 38 5f 42 4f
0000020 4d 0d 0a e6 b5 8b e8 af 95 0d 0a e4 b8 ad e6 96
0000040 87
0000041
UTF8_noBOM.java
0000000 74 65 73 74 0d 0a 55 54 46 5f 6e 6f 42 4f 4d 2e
0000020 6a 61 76 61 0d 0a e6 b5 8b e8 af 95 0d 0a e4 b8
0000040 ad e6 96 87
0000044
```
